src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -394,6 +394,29 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
 
 				AppendShardIdToName(oldRelationName, shardId);
 				AppendShardIdToName(newRelationName, shardId);
+
+				/*
+				 * PostgreSQL creates array types for each ordinary table, with
+				 * the same name plus a prefix of '_'.
+				 *
+				 * ALTER TABLE ... RENAME TO ... also renames the underlying
+				 * array type, and the DDL is run in parallel connections over
+				 * all the placements and shards at once. Concurrent access
+				 * here deadlocks.
+				 *
+				 * Let's provide an easier to understand error message here
+				 * than the deadlock one.
+				 *
+				 * See also https://github.com/citusdata/citus/issues/1664
+				 */
+				if (strlen(*newRelationName) >= (NAMEDATALEN - 1))
```

Comment:
A minor Citus convention (which I think derives from Code Complete 2) is to avoid function calls in expressions where possible. So the preferred pattern is:

```
newRelationNameLength = strlen(*newRelationName);
if (newRelationNameLength >= (NAMEDATALEN - 1))
{
   ...
}
```

That way we could set a break point on the `if` and see the actual length that causes the expression to evaluate to true.